guardian/cdk
# feat!: Remove `Stage` CFN Parameter in favour of property

The diff on this PR is pretty big! Sadly it's unavoidable as we're removing a property from a class that's used throughout the repository.

It might be easier to review commit by commit? I plan to squash the commits at merge time to keep the release notes clean.

## What does this change?


Following https://github.com/guardian/cdk/pull/1123, this change removes all stage awareness from GuCDK in favour of producing a template per stage. This means we no longer produce [Mappings](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/mappings-section-structure.html) in the template, and users can make _any_ property stage aware.

This change works in conjunction with https://github.com/guardian/riff-raff/pull/681.

BREAKING CHANGE: Removal of explicit stage awareness in favour of users defined properties.

From:

```ts
// -- lib/cdk-playground.ts
export class CdkPlayground extends GuStack {
  private static app = "cdk-playground";

  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    new GuPlayApp(this, {
      app: CdkPlayground.app,
      instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.MICRO),
      access: { scope: AccessScope.PUBLIC },
      userData: {
        distributable: {
          fileName: `${CdkPlayground.app}.deb`,
          executionStatement: `dpkg -i /${CdkPlayground.app}/${CdkPlayground.app}.deb`,
        },
      },
      certificateProps: {
        [Stage.CODE]: {
          domainName: "cdk-playground.code.dev-gutools.co.uk",
        },
        [Stage.PROD]: {
          domainName: "cdk-playground.prod.gutools.co.uk",
        },
      },
      monitoringConfiguration: { noMonitoring: true },
      scaling: {
        [Stage.CODE]: {
          minimumInstances: 1,
        },
        [Stage.PROD]: {
          minimumInstances: 3,
        },
      },
    });
  }
}

// -- bin/cdk.ts
const app = new App();

// produces one output file `cdk.out/CdkPlayground.template.json` with `Mappings` and `FindInMap` intrinsic functions
new CdkPlayground(app, "CdkPlayground", { stack: "deploy" });
```

To:

```ts
// -- lib/cdk-playground.ts
export interface CdkPlaygroundProps {
  stage: string;
  domainName: string;
  scaling: GuAsgCapacity;
}

export class CdkPlayground extends GuStack {
  private static app = "cdk-playground";

  constructor(scope: App, id: string, props: CdkPlaygroundProps) {
    const { stage, domainName, scaling } = props;

    super(scope, id, {
      stack: "deploy",
      stage,
    });

    new GuPlayApp(this, {
      app: CdkPlayground.app,
      instanceType: InstanceType.of(InstanceClass.T4G, InstanceSize.MICRO),
      access: { scope: AccessScope.PUBLIC },
      userData: {
        distributable: {
          fileName: `${CdkPlayground.app}.deb`,
          executionStatement: `dpkg -i /${CdkPlayground.app}/${CdkPlayground.app}.deb`,
        },
      },
      certificateProps: {
        domainName,
      },
      monitoringConfiguration: { noMonitoring: true },
      scaling,
    });
  }
}

// -- bin/cdk.ts
const app = new App();

// produces output file `cdk.out/CdkPlayground-CODE.template.json`
new CdkPlayground(app, "CdkPlayground-CODE", {
  stage: "CODE",
  domainName: "cdk-playground.code.dev-gutools.co.uk",
  scaling: { minimumInstances: 1 },
});

// produces output file `cdk.out/CdkPlayground-PROD.template.json`
new CdkPlayground(app, "CdkPlayground-PROD", {
  stage: "PROD",
  domainName: "cdk-playground.prod.gutools.co.uk",
  scaling: { minimumInstances: 3 },
});
```

## How to test



- Install GuCDK from this branch into a project
- Update the stack definition of the project
- Update the `riff-raff.yaml` of the project, adding [`templateStagePaths`](https://github.com/guardian/riff-raff/pull/681)
- Deploy the project, observe the correct template file being used 

Indeed this has been done on cdk-playground [here](https://github.com/guardian/cdk-playground/pull/173).

## How can we measure success?


- Any property of a stack can be made stage aware at the user's choosing
- A much simplified code base

## Have we considered potential risks?


This is quite the breaking change! The release notes should be simple and easy to follow!

❓ The migration guides will be out of date after this change. Can we update them _after_ this has gone out and we've migrated one of our YAML defined stacks?

## Checklist


[^1]: Consider whether this is something that will mean changes to projects that have already been migrated, or to the CDK CLI tool. If changes are required, consider adding a checklist here and/or linking to related PRs.
[^2]: If you are adding a new construct or pattern, has new documentation been added? If you are amending defaults or changing behaviour, are the existing docs still valid?
